ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/FilterStringColLikeStringScalar.java

```diff
@@ -51,93 +53,83 @@ public FilterStringColLikeStringScalar(int colNum, byte[] likePattern) {
-    }
-  }
+    },
+    // Accepts any LIKE patterns and creates corresponding checkers.
+    COMPLEX(ComplexChecker.class) {
+      @Override
+      public String format(String pattern) {
+        return "^" + UDFLike.likePatternToRegExp(pattern) + "$";
+      }
+    },
+    // Accepts chained LIKE patterns without escaping like "abc%def%ghi%" and
+    // creates corresponding checkers.
+    CHAINED(ChainedChecker.class),
+    // Accepts simple LIKE patterns like "abc" and creates corresponding checkers.
+    NONE(NoneChecker.class);
 
-  /**
-   * Accepts simple LIKE patterns like "abc" and creates corresponding checkers.
-   */
-  private static class NoneCheckerFactory implements CheckerFactory {
-    private static final Pattern NONE_PATTERN = Pattern.compile("[^%_]+");
+    Class<? extends Checker> checker;
 
-    public Checker tryCreate(String pattern) {
-      Matcher matcher = NONE_PATTERN.matcher(pattern);
-      if (matcher.matches()) {
-        return new NoneChecker(pattern);
-      }
-      return null;
+    UDFLikePattern(Class<? extends Checker> checker) {
+      this.checker = checker;
     }
-  }
 
-  /**
-   * Accepts chained LIKE patterns without escaping like "abc%def%ghi%" and creates corresponding
-   * checkers.
-   *
-   */
-  private static class ChainedCheckerFactory implements CheckerFactory {
-    private static final Pattern CHAIN_PATTERN = Pattern.compile("(%?[^%_\\\\]+%?)+");
-
-    public Checker tryCreate(String pattern) {
-      Matcher matcher = CHAIN_PATTERN.matcher(pattern);
-      if (matcher.matches()) {
-        return new ChainedChecker(pattern);
+    private static UDFLikePattern matcher(String pattern) {
+      UDFLikePattern lastType = NONE;
+      int length = pattern.length();
+      char lastChar = 0;
+
+      for (int i = 0; i < length; i++) {
+        char n = pattern.charAt(i);
+        if (n == '_' && lastChar != '\\') { // such as "a_bc"
+          return COMPLEX;
+        } else if (n == '%') {
+          if (i == 0) { // such as "%abc"
+            lastType = END;
+          } else if (i < length - 1) {
+            if (lastChar != '\\') { // such as "a%bc"
+              lastType = CHAINED;
+            }
+          } else {
+            if (lastChar != '\\') {
+              if (lastType == END) { // such as "%abc%"
+                lastType = MIDDLE;
+              } else if (lastType != CHAINED) {
+                lastType = BEGIN; // such as "abc%"
+              }
+            }
+          }
+        }
+        lastChar = n;
       }
-      return null;
+      return lastType;
     }
-  }
 
-  /**
-   * Accepts any LIKE patterns and creates corresponding checkers.
-   */
-  private static class ComplexCheckerFactory implements CheckerFactory {
-    public Checker tryCreate(String pattern) {
-      // anchor the pattern to the start:end of the whole string.
-      return new ComplexChecker("^" + UDFLike.likePatternToRegExp(pattern) + "$");
+    public String format(String pattern) {
```

Comment:
no need for `public`, package-private is enough

Reply:
I will fix. Then run the test again.